sycl/doc/syclcompat/README.md

```diff
@@ -0,0 +1,1288 @@
+                         size_t x, size_t y,
+                         sycl::queue q = get_default_queue()); // 2D matrix
+sycl::event memset_async(pitched_data pitch, int val,
+                         sycl::range<3> size,
+                         sycl::queue q = get_default_queue()); // 3D matrix
+
+void free(void *ptr, sycl::queue q = get_default_queue());
+sycl::event free_async(const std::vector<void *> &pointers,
+                       const std::vector<sycl::event> &events,
+                       sycl::queue q = get_default_queue());
+
+/// Queries pointer allocation type
+class pointer_attributes {
+public:
+  void init(const void *ptr, sycl::queue q = get_default_queue());
+  sycl::usm::alloc get_memory_type();
+  const void *get_device_pointer();
+  const void *get_host_pointer();
+  bool is_memory_shared();
+  unsigned int get_device_id();
+};
+
+} // syclcompat
+```
+
+Finally, the class `pitched_data`, which manages memory allocation for 3D
+spaces, padded to avoid uncoalesced memory accesses.
+
+```c++
+namespace syclcompat {
+
+class pitched_data {
+public:
+  pitched_data();
+  pitched_data(void *data, size_t pitch, size_t x, size_t y);
+
+  void *get_data_ptr();
+  size_t get_pitch();
+  size_t get_x();
+  size_t get_y();
+
+  void set_data_ptr(void *data);
+  void set_pitch(size_t pitch);
+  void set_x(size_t x);
+  void set_y(size_t y);
+};
+
+} // syclcompat
+```
+
+There are various helper classes and aliases defined within SYCLcompat to
+encapsulate and define memory operations and objects. These classes and aliases
+are primarily designed to assist with machine translation from other
+heterogeneous programming models.
+
+The wrapper class `device_memory` provides a unified representation for device
+memory in various regions. The class provides methods to allocate memory for the
+object (`init()`) and access the underlying memory in various ways (`get_ptr()`,
+`get_access()`, `operator[]`). Aliases for global and USM shared specializations
+are provided.
+
+The `memory_traits` class is provided as a traits helper for `device_memory`.
+The `accessor` class template provides a 2D or 3D `sycl::accessor`-like wrapper
+around raw pointers.
+
+```c++
+namespace syclcompat {
+
+enum class memory_region {
+  global = 0, // device global memory
+  local,      // device local memory
+  usm_shared, // memory which can be accessed by host and device
+};
+
+using byte_t = uint8_t;
+
+enum class target { device, local };
+
+template <memory_region Memory, class T = byte_t> class memory_traits {
+public:
+  static constexpr sycl::access::address_space asp =
+      (Memory == memory_region::local)
+          ? sycl::access::address_space::local_space
+          : sycl::access::address_space::global_space;
+  static constexpr target target =
+      (Memory == memory_region::local)
+          ? target::local
+          : target::device;
+  static constexpr sycl::access_mode mode = sycl::access_mode::read_write;
```

Comment:
Shall we add another memory_region enum `constant`, and set `mode = sycl::access_mode::read` if `Memory` is `constant` ?

Reply:
We modified this from the original DPCT helper header code to reflect the fact that SYCL's `constant_space` is deprecated in SYCL 2020. It depends how closely `memory_region` is aligned with SYCL's `access::address_space`. At present we have a `usm_shared` option for the `memory_region` enum, so it doesn't *exactly* reflect the `address_space`.

I can see two scenarios in which we would need/want to have a `memory_region::constant` option:
 1. We intentionally continue to support the deprecated `address_space`
 2. We have some other notion of `memory_region::constant` like "it's actually `global` but we want to use it read-only".

Do you feel that one of these situations applies here?

Reply:
> We have some other notion of memory_region::constant like "it's actually global but we want to use it read-only".
For SYCLomatic, it can know the original memory is qualified with `__constant__`, so we can set `sycl::access_mode::read` for it to hint compiler to do more optimizations. That is the reason it want to keep `memory_region::constant`

Reply:
Yes that seems reasonable, we can add it back in.

Reply:
@zhiweij1, We've added the constant back. Let us know if there is something else before we proceed with the merge. Thanks!

Reply:
I'm OK with this PR. @AndyCHHuang @tomflinda @ziranzha @zhimingwang36 , do you have other comments?

Reply:
Nice. Since it's been quite some time, I'm gonna resolve this, but feel free to reach us through other means (issues,  mail) if we need to address something else. Thanks